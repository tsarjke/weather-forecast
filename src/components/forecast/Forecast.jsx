import React, { useState } from 'react';
import getWeatherData from '../../services/getWeatherData';
import getForecastData from '../../utils/getForecastData';
import Loading from '../loading/Loading';
import Error from '../error/Error';
import Cards from '../cards/Cards';
import './forecast.scss';
import SearchPanel from '../searchPanel/SearchPanel';
import Switcher from '../switcher/Switcher';

const Forecast = () => {
  const [searchText, setSearchText] = useState('');
  const [measure, setMeasure] = useState('celsius');
  const [weatherData, setWeatherData] = useState([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);

  const onFormSubmit = async (e) => {
    e.preventDefault();

    if (searchText) {
      try {
        setError(false);
        setLoading(true);

        const answer = await getWeatherData(searchText);
        if (answer.forecast) {
          const {
            forecast: { forecastday },
          } = answer;
          setWeatherData(getForecastData(forecastday));
        } else if (answer.error) {
          setError('City not found');
        } else {
          setError(
            'It is impossible to get the data, contact the copyright holder',
          );
        }
      } catch {
        setError('Something went wrong');
      } finally {
        setLoading(false);
      }
    } else {
      setError(true);
    }
  };

  let content;

  const handleChange = ({ target }) => {
    setError(false);

    if (target.type === 'text') {
      setSearchText(target.value);
      if (weatherData.length > 0) {
        setWeatherData([]);
      }
    }
    if (target.type === 'radio') {
      setMeasure(target.value);
    }
  };

  if (error) {
    content =
      searchText === '' ? (
        <Error text="Enter the name of the locality" />
      ) : (
        <Error text={error} />
      );
  } else if (loading) {
    content = <Loading />;
  } else {
    content =
      weatherData.length === 0 ? (
        ''
      ) : (
        <Cards
          data={weatherData}
          unit={measure}
        />
      );
  }

  return (
    <div
      className="forecast"
      data-testid="forecast"
    >
      <div className="forecast__content">
        <form
          onSubmit={onFormSubmit}
          action="#"
          className="forecast__form"
          data-testid="forecast-form"
        >
          <SearchPanel
            handleChange={handleChange}
            value={searchText}
          />
          <Switcher
            handleChange={handleChange}
            value={measure}
          />
        </form>
        <div
          className="forecast__content"
          data-testid="content"
        >
          {content}
        </div>
      </div>
    </div>
  );
};
export default Forecast;
