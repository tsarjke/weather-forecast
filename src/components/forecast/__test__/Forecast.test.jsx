import { fireEvent, render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import Forecast from '../Forecast';
import mockGetWeatherData from '../../../services/getWeatherData';

jest.mock('../../../services/getWeatherData.js');

describe('Forecast', () => {
  // beforeEach(() => {
  //   mockGetWeatherData.mockResolvedValue({ cod: '404' });
  // });

  test('should render the Error component and ask to write comething in search input as search input is empty', () => {
    render(<Forecast />);

    const forecastForm = screen.getByTestId('forecast-form');
    fireEvent.submit(forecastForm);
    expect(
      screen.getByText('Enter the name of the locality'),
    ).toBeInTheDocument();
  });

  test('Should show the Error component with "City not found" when the 404 cod is received', async () => {
    mockGetWeatherData.mockResolvedValue({ cod: '404' });
    render(<Forecast />);

    const searchInput = screen.getByTestId('search-input');
    fireEvent.change(searchInput, { target: { value: 'San-francisco' } });

    await act(async () => {
      const forecastForm = screen.getByTestId('forecast-form');
      fireEvent.submit(forecastForm);
    });

    expect(screen.getByText('City not found')).toBeInTheDocument();
    expect(mockGetWeatherData).toHaveBeenCalledTimes(1);
  });

  test('Should show the Error component with "Something went wrong" when a rejection is received', async () => {
    mockGetWeatherData.mockRejectedValue();
    render(<Forecast />);

    const searchInput = screen.getByTestId('search-input');
    fireEvent.change(searchInput, { target: { value: 'San-francisco' } });

    await act(async () => {
      const forecastForm = screen.getByTestId('forecast-form');
      fireEvent.submit(forecastForm);
    });

    expect(screen.getByText('Something went wrong')).toBeInTheDocument();
    expect(mockGetWeatherData).toHaveBeenCalledTimes(1);
  });

  test('should render form and empty div for content', () => {
    render(<Forecast />);

    const forecastElem = screen.getByTestId('forecast');
    expect(forecastElem).toMatchSnapshot();
  });
});
