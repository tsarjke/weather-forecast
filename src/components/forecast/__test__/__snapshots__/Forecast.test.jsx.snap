// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forecast should render form and empty div for content 1`] = `
<div
  class="forecast"
  data-testid="forecast"
>
  <div
    class="forecast__content"
  >
    <form
      action="#"
      class="forecast__form"
      data-testid="forecast-form"
    >
      <div
        class="search"
      >
        <input
          class="search__text-input"
          data-testid="search-input"
          name="city"
          type="text"
          value=""
        />
        <button
          class="search__btn"
          data-testid="search-btn"
          type="submit"
        >
          Search
        </button>
      </div>
      <div
        class="switcher"
      >
        <div
          class="switcher__content"
        >
          <label
            class="switcher__label"
            for="celsius"
          >
            <input
              checked=""
              class="switcher__radio"
              data-testid="celsius-btn"
              id="celsius"
              name="measure"
              type="radio"
              value="celsius"
            />
            <span>
              Celsius
            </span>
          </label>
          <label
            class="switcher__label"
            for="fahrenheit"
          >
            <input
              class="switcher__radio"
              data-testid="fahrenheit-btn"
              id="fahrenheit"
              name="measure"
              type="radio"
              value="fahrenheit"
            />
            <span>
              Fahrenheit
            </span>
          </label>
        </div>
      </div>
    </form>
    <div
      class="forecast__content"
      data-testid="content"
    />
  </div>
</div>
`;
